### PR TITLE
chore: Remove delay starting the change detection service

### PR DESCRIPTION
**Pull request by polarathene**
# Description

There does not seem to be a need anymore for delaying the service availability by 10 seconds when starting a container.
- A test needed to be adjusted as Amavis was not ready to receive emails without the extra time (delay) that the change detection service provided. This is not a bug with the change detector :)
- The change event polling now only replaces the existing checksum when the contents has actually changed (_after processing_). I don't expect that to cause regressions, but it might reduce some redundant I/O each poll?
- The change service is now the last daemon started.

AFAIK, with the current start-up scripts the change detector service should not have an issue with conflicting writes:
- It would be started after all our setup scripts have run (initial checksum created), followed by permission fixes, user-patches and other daemons being started.
- For first runs without any accounts, this occurs during setup scripts, before the change detector service is running, so again I don't see any race conditions?
- This was apparently a concern / issue in the past. One concern raised was multiple containers sharing a config volume (_such as over a network NFS share_), I'm not sure that a fixed sleep would alleviate that though?

## History + References

- When the [change detection feature was originally contributed in Oct 2017](https://github.com/docker-mailserver/docker-mailserver/pull/738) (_[associated issue](https://github.com/docker-mailserver/docker-mailserver/issues/552)_), the script would begin with `sleep 5`.
- In [July 2019, that was changed to `sleep 10`](https://github.com/docker-mailserver/docker-mailserver/pull/1201) and re-located to sleep at a later point.
- For reference, earlier PRs I worked on that adjusted change detection start-up handling and provide history around that: https://github.com/docker-mailserver/docker-mailserver/pull/2610 + https://github.com/docker-mailserver/docker-mailserver/pull/2623

<details>
<summary>Change Detector start-up flow (preparing initial checksum)</summary>

`_setup` => `_prepare_for_change_detection` => `_monitored_files_checksums >"${CHKSUM_FILE}" `

https://github.com/docker-mailserver/docker-mailserver/blob/00b1d88ed74e2cc90f70dede4e37a5ce9d1fdc6a/target/scripts/start-mailserver.sh#L228-L242

https://github.com/docker-mailserver/docker-mailserver/blob/00b1d88ed74e2cc90f70dede4e37a5ce9d1fdc6a/target/scripts/startup/setup-stack.sh#L3-L13

https://github.com/docker-mailserver/docker-mailserver/blob/00b1d88ed74e2cc90f70dede4e37a5ce9d1fdc6a/target/scripts/helpers/change-detection.sh#L9-L21

</details>

## Type of change

- [x] Improvement (non-breaking change that does improve existing functionality)

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] If necessary I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Review by georglauterbach — APPROVED** [review]
I re-ran the changes as we were seeing weird issues with `smtp_delivery.bats`. Looks fine now though. You could do me the favour and run the tests locally a few time so we can be more sure about not having introduced a new race condition here too.

File: test/tests/parallel/set3/mta/smtp_delivery.bats
Comment (polarathene) [reply test/tests/parallel/set3/mta/smtp_delivery.bats:55]: I'm not actually sure when these are used :sweat_smile: 

Perhaps I'm listening on the wrong port since 10024 appears to be for sending and 10025 for receiving? I just know that when I queried the mail queue, sometimes a connection refused error mentioned port 10024.

> You could do me a favour and run the tests locally a few times so we can be more sure about not having introduced a new race condition here.

I ran it many times actually and wasn't able to produce it locally. I had this change ready around a month ago, but back then it occurred more often. I had a solution to delay sending mail a little bit with `sleep 1`:

```bash
  # NOTE: If change detection begins too early, it can cause connection issues with Amavis.
  # `sleep 1` can be added before or after to prevent that from happening.
  sleep 1
```

I am pretty sure that the issue is more to do with Amavis not being ready quick enough, as before with the 10 sec change event delay, there was plenty of time. There might be a small window after the change event with the reloads on amavis and postfix (plus dovecot for delivery I guess).

---

I'll be getting some rest, but will address this concern when I'm back :+1:

Comment (georglauterbach) [reply test/tests/parallel/set3/mta/smtp_delivery.bats:55]: > I'll be getting some rest, but will address this concern when I'm back 👍

Alright! I too think this is about a Amavis-timing-issue. I'd add the check for port 10025 and I'm not against a small sleep as mentioned in your comment (with additional info on why it's needed, just like in your comment).

Comment (casperklein) [reply test/tests/parallel/set3/mta/smtp_delivery.bats:55]: >  I'd add the check for port 10025

Will this be added or should I approve now?

Comment (polarathene) [reply test/tests/parallel/set3/mta/smtp_delivery.bats:55]: I'll update it, sorry been a bit busy and since CI is broken presently for tests it seems I should wait until we've resolved that :sweat_smile:

Comment (casperklein) [reply test/tests/parallel/set3/mta/smtp_delivery.bats:55]: > I'd add the check for port 10025

Will this be added or should I approve now? 😆

Comment (polarathene) [reply test/tests/parallel/set3/mta/smtp_delivery.bats:55]: Oh sorry, I don't believe there needs to be a check for port 10025. AFAIK this is a port on Postfix to receive mail back from Amavis after it's done processing. I am doubtful it'll make a difference when there's a connection error logged about port 10024. The sleep should give Amavis enough time.

**Comment by polarathene** [comment]
@casperklein ready to approve :sweat_smile:

**Comment by georglauterbach** [comment]
> @casperklein ready to approve :sweat_smile: 

If you're really short on time, just let us briefly know :) Maybe we can then go ahead and merge a few PRs with only reviewer :)

**Review by georglauterbach — APPROVED** [review]

**Comment by georglauterbach** [comment]
🤦🏼 I did not see auto-merge was enabled...

The merge happened because I changed the required reviewers for the time being to 1. Rationale: This project is maintained by volunteers, and we cannot ask people to spend time on it when they do not have time for it. On the other hand, we have at least 1 reviewer available at the moment, so I think _for the time being_, it is safe to merge some PRs with only 1 approval. Best case would be to have 2 frequent reviewers again. @casperklein just keep us updated on your current status, I can change the required approvals back at any time :)

I also wanted to ask @polarathene if you're fine with this being merged before actually merging it, but the auto-merge was just too fast.. :D

**Comment by polarathene** [comment]
> I changed the required reviewers for the time being to 1. Rationale: This project is maintained by volunteers, and we cannot ask people to spend time on it when they do not have time for it.
> 
> On the other hand, we have at least 1 reviewer available at the moment, so I think _for the time being_, it is safe to merge some PRs with only 1 approval, especially when PRs are waiting for a longer time to get merged (hindering more work to happen ATM).

I think that this is the right thing to do.

We have not been getting new maintainers joining, and many of us are strained for time. I know I've over-committed myself already and need to focus on other tasks outside of DMS.

As much as I'd like to contribute more, it's a huge time sink that's difficult to justify the amount of free time I invest into it :sweat_smile:

**Comment by georglauterbach** [comment]
No worries, you have already contributed so much and you added so much value to this project ❤️ I know from experience that maintaing takes a lot of time, absolutely right :D I think what would be most valuable is to still have a few people around for reviewing some of the proposed changes - one does not need to invest heavily and doing changes themselves; reviewing is very important in the end :)

**Comment by casperklein** [comment]
> @casperklein just keep us updated on your current status

I am still here. I just had some two extreme busy weeks and had no time for reviews. If PRs are not related to the current test improvements, I'll review them in more depth. So I don't approve them in hurry to get them merged asap (without reviewing them properly). I will take a look at the merged PRs, in case something important was missed.

**Comment by georglauterbach** [comment]
> > @casperklein just keep us updated on your current status
> 
> I am still here. I just had some two extreme busy weeks and had no time for reviews.

I hope I did not offend you in any way :) I am going to be on vacation and I'll be very busy the next weeks, so I wanted to get some PRs out of the way, to have peace of mind during my vacation :D

> If PRs are not related to the current test improvements, I'll review them in more depth. So I don't approve them in hurry to get them merged asap (without reviewing them properly). I will take a look at the merged PRs, in case something important was missed.

👍🏼

**Comment by casperklein** [comment]
> I hope I did not offend you in any way :)

No you didn't. In march I'll be more active again.

**Pull request merged**
